src/include/storage/index/hash_index.h

```diff
@@ -0,0 +1,272 @@
+#pragma once
+
+#include <functional>
+#include <unordered_set>
+#include <utility>
+#include <variant>  // NOLINT (Matt): lint thinks this C++17 header is a C header because it only knows C++11
+#include <vector>
+#include "farmhash/farmhash.h"
+#include "libcuckoo/cuckoohash_map.hh"
+#include "storage/index/index.h"
+#include "transaction/transaction_context.h"
+#include "transaction/transaction_manager.h"
+
+namespace terrier::storage::index {
+
+#define INITIAL_CUCKOOHASH_MAP_SIZE 256
+
+/**
+ * Wrapper around libcuckoo's hash map.
+ * @tparam KeyType the type of keys stored in the map
+ */
+template <typename KeyType>
+class HashIndex final : public Index {
+  friend class IndexBuilder;
+
+ private:
+  struct TupleSlotHash {
+    std::size_t operator()(const TupleSlot &slot) const {
+      return util::Hash64(reinterpret_cast<const char *>(&slot), sizeof(TupleSlot));
+    }
+  };
+
+  using ValueMap = std::unordered_set<TupleSlot, TupleSlotHash>;
+  using ValueType = std::variant<TupleSlot, ValueMap>;
+
+  HashIndex(const catalog::index_oid_t oid, const ConstraintType constraint_type, IndexMetadata metadata)
+      : Index(constraint_type, std::move(metadata)),
+        hash_map_{new cuckoohash_map<KeyType, ValueType>(INITIAL_CUCKOOHASH_MAP_SIZE)} {}
+
+  cuckoohash_map<KeyType, ValueType> *const hash_map_;
+
+ public:
+  ~HashIndex() final { delete hash_map_; }
+
+  void PerformGarbageCollection() final{};
+
+  bool Insert(transaction::TransactionContext *const txn, const ProjectedRow &tuple, const TupleSlot location) final {
+    TERRIER_ASSERT(GetConstraintType() == ConstraintType::DEFAULT,
+                   "This Insert is designed for secondary indexes with no uniqueness constraints.");
+    KeyType index_key;
+    index_key.SetFromProjectedRow(tuple, metadata_);
+
+    bool UNUSED_ATTRIBUTE insert_result = false;
+
+    auto key_found_fn = [location, &insert_result](ValueType &value) -> bool {
+      if (std::holds_alternative<TupleSlot>(value)) {
+        // replace it with a ValueMap
+        const auto existing_location = std::get<TupleSlot>(value);
+        value = ValueMap({{location}, {existing_location}}, 2);
+        insert_result = true;
+      } else {
+        // insert the location to the cuckoohash_map
+        auto &value_map = std::get<ValueMap>(value);
+        insert_result = value_map.emplace(location).second;
+      }
+      return false;
+    };
+
+    const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, key_found_fn, location);
+
+    TERRIER_ASSERT(insert_result != uprase_result,
+                   "Either a new key was inserted (uprase_result), or the value already existed and a new value was "
+                   "inserted (insert_result).");
+
+    // Register an abort action with the txn context in case of rollback
+    txn->RegisterAbortAction([=]() {
+      auto abort_key_found_fn = [location](ValueType &value) -> bool {
```

Comment:
A doxygen-like comment should precede this given the general complexity of this abort action (defering a capturing labmda that passes a capturing lambda to a third-party library).